Feature mud element

porkopek opened the pull request:
Several features in several commits that can be cherry picked.

 Added : Primitive component `MudElement` that allows to control the markup of a given component though the HtmlTag property. This enables several improvements, like render a MudButton as an anchor element, so there is no need in using JS to navigate.
Added: Example of using HtmlTag property in a Link Button
Added: Examples of composing a File Uploader Button

Changed: all buttons inner code, to take advantage of the new MudElement component

This is a draft that will be upgraded with your comments

henon commented:
I will resolve the conflicts later.  This is how the first example looks:

![image](https://user-images.githubusercontent.com/44090/101666484-8ea57400-3a4e-11eb-813f-4fe635fc4c39.png)

And the example code.

```html
@using BlazorInputFile
<MudButton HtmlTag="label"
           Variant="Variant.Filled"
           Color="Color.Primary"           
           StartIcon="@Filled.CloudUpload"
           for="fileInput">
    Upload Files
</MudButton>

<InputFile id="fileInput" OnChange="UploadFiles" hidden multiple/>

@if (files != null)
{
    <MudList>
        @foreach (var file in files)
        {
            <MudListItem Icon="@Filled.AttachFile">
                @file.Name <code>@file.Size bytes</code>
            </MudListItem>
        }
    </MudList>
}
@code{
    IEnumerable<IFileListEntry> files;
    private void UploadFiles(IFileListEntry[] entries)
    {
        files = entries;
        //TODO upload the files to the server
    }
}
```

Now I have a question regarding this because I am puzzled by the text. It says that on 3.1 I need the nuget. But I am on 3.1 here and just selected a file and it works fine. The text suggests, that `<InputFile>` only works on net5. What did I miss?

henon commented:
I am stupid. On merging the csproj files I see that you added the nuget ;)

`<PackageReference Include="BlazorInputFile" Version="0.2.0" />`

henon commented:
ok, I would say we make the file input available only for net5. 
Supporting new features in 3.1 is not really what we want anymore. We have to support 3.1 for some who are stuck with it but 99% of people are using net5 anyway. So I guess we just change the text to "net5 required for this feature"

porkopek commented:
We don't need to support FileInput if we limit this to what really is, an example of an implementation. In this example, I show how to implement the File Upload Button, and it is very straight forward. I think this isn't a component, because you have more freedom composing it as you want.  It Is just 2 components working together: a MudButton working as a label and a FileInput (or whatever implementation the user chooses). You don't win almost anything encapsulating it as a single component. There is no benefit, but there are disadvantages: you only can use that component in 5.0.
But making it the way I show you here, you can use it in 3.1 or 5.0, it's the user's choice. If she is going to upload a file, she is going to need a FileInput component or her own implementation anyway, so the only thing that she really needs it's just a button that renders a label.
This way you give absolute freedom to the user.
And this is just one example of the power of MudElement. MudElement enables the rendering of whatever html tag you want

porkopek commented:
I'll provide more examples, as a drag and drop file uploader

henon commented:
This is so awesome

![image](https://user-images.githubusercontent.com/44090/101674485-dcbf7500-3a58-11eb-96ef-be1967bd415c.png)

Thanks to MudElement no JS is required any more for that.

henon commented:
```c#
        /// <summary>
        /// The HTML element that will be rendered in the root by the component
        /// </summary>
        [Parameter] public string HtmlTag { get; set; }
```

I would not add this to MudComponentBase. For many components it doesn't make sense to have this parameter. I'd rather add it to the components where it makes real functional sense for the user to change it.

porkopek commented:
We can start by adding it just to the button components, but I think that every component should benefit of being able to choose its root markup. I also should add some tests, when I learn how to use BUnit.
What do you think about what I was telling you on file uploader? Do you still want to encapsulate it and limit it only just for 5.0? I don't see any real benefits

porkopek commented:
Maybe we can start just adding the file upload example and in a subsequent release encapsulate it, if we see that it makes sense

porkopek commented:
The conflicts regarding MudButon and MudFab are because MudButton, in the last implementation added `onclick:stopPropagation="true"`  (don't know why) and in my implementation this is not added. I think that StopPropagation must be a property, to optional allow to the user if he wants to stop propagation or not. I can add it if you want.

porkopek commented:
Changes that I've implemented:
1) HtmlTag is not a property of MudComponentBase, but only of MudBaseButton, like discussed
2) I've updated the file upload example to look nicer and more clear
3) I've changed one line of MudFab className to apply  missing styles to allow to convert buttons in another html elements and don't miss styles. Related to  #286 
4) I've created the MudElement documentation page with examples
5) I've added tests for MudButton that renders an anchor and for MudElement
6) I've removed two obsolete properties from MudButton that were used to navigate using JS, because with MudElement rendering an anchor are no longer necessary

In subsequent improvements, I'll add a drag and drop file upload example and maybe, if we think it's cool, a file upload button encapsulation

henon commented:
1) Make sure to discuss this PR with @Garderoben as I know there is some resistance on his behalf regarding the MudButton changes. I think Garderoben has not yet understood the advantages of your solution.

2) The Stoppropagation is needed to be true per default or it will lead to problems in other mud components which use buttons. If someone really needs click propagation they can make it a parameter, but I can't think of a use case.

porkopek commented:
Yeah, I completely get it. This PR is very hard to analyze, because has a lot of changes, some of them structural.
If I were you, I would be afraid to accept it.
So, on second thought, I'd better take it down and maybe I'll PR just the MudElement, so that can be used both internally and externally
